### DISCUSSION TITLE
Welcome to OpenTK Discussions!

### DISCUSSION THREAD

**Discussion by NogginBops** [opentk/opentk#1492] [General]
## Welcome!
I've decided that it's time to open up GitHub discussions!
My aim with discussions is to open up a forum-like place to be able to have focused discussions about OpenTK that are easily linked to and recalled. 
A tool for documenting intent of sorts.

Any questions or discussions related to OpenTK are allowed here, although for support questions our discord server will continue to be the main place to get help. 

Join the discord here!
[![Discord](https://discordapp.com/api/guilds/337627185248468993/widget.png)](https://discord.gg/6HqD48s)

This post is meant to kickstart discussions here on GitHub, to join the conversation please comment below with an introduction! 👋 

## Hello! 👋

Hello! My name is @NogginBops  and I'm currently the head maintainer of OpenTK! 
I've been maintainer for almost 2 years as of writing this. 
It's been a pleasure to continue the long legacy of OpenTK, a project which is more than 16 years since it's initial commit (and roots that go even further back in time).

I've been involved in basically every part of OpenTK, which contains a surprisingly large number of projects.
Answering questions on discord, merging fixes into OpenTK 3, continuing to fix and add features to OpenTK 4, developing a new bindings generator with @frederikja163 for OpenTK 5, implementing a win32 backend for the PAL2 project (OpenTK 5+), maintaining the GLControl and GLWpfControls, updating the website and I'm planning on rewriting the LearnOpenTK tutorials whenever I get some time to spare. 

Please share something about yourself too! It doesn't have to be OpenTK related! 
Tell us about what kinds of projects you are working on (or what you want to work on).

**Comment by utkumaden** [comment]
Hello, it's me themixedupstuff. At the time of writing I'm a senior bachelor's student of electrical and electronics engineering. I'm also currently the lead of the PAL 2.0 project and have been painstakingly working on a Linux backend for it. Due to my very busy personal schedule things can move a bit slow but I'd love to see new faces on the library as a whole who are willing to help.

Usually you will see me around in the server in #support and often be the first person you interact with. I'll try my best to answer your questions about the library and writing OpenGL. I am also one of the moderators of the discord server so if you happen to see something fishy feel free to ping me. Got any electronics or embedded questions too? I can answer them in #general if it's not busy.

See you all around 👋

**Comment by NotSignificant** [comment]
Hi there
I recognized a problem starting with a fullscreen window and then switch back to a normal window. I always get an exception, saying that 0x0 is an invalid window size.
If I start with a normal window and switch to fullscreen and back this doesn't happen. It seems, _cachedWindowClientSize and _cachedWindowLocation will not be initialized, when starting in fullscreen. They should be set to default values (i.e. 640x480 and 0x0), I suggest.

**Reply by NogginBops** [reply]
Hello!
Could you open a proper issue for this?
https://github.com/opentk/opentk/issues/new

**Reply by NotSignificant** [reply]
Done.

**Comment by BuslikDrev** [comment]
На сколько OpenTK 3 медленнее OpenTK 4-5, и какого важного функционала не имеет?

**Reply by NogginBops** [reply]
OpenTK 3 runs on .net framework which doesn't have some of the runtime optimizations of later .net versions.
Some of the math functions are more optimized in OpenTK 4.
Other than that I don't expect there to be any significant difference in performance between OpenTK 3 and OpenTK 4.

**Reply by BuslikDrev** [reply]
Провёл сравнение https://www.shadertoy.com/view/4dBcRD:
OpenGL Chrome GPU 50% 60 fps
OpenTK 3.3.3 WinForms GPU 100% 40 fps

Возможно потому-что браузер использует Vulkan.

**Reply by NogginBops** [reply]
This is completely dependent on how you've implemented this in OpenTK, and also probably *very* resolution dependent.>
Check to see if the OpenTK window has a higher resolution than what is displayed in the web window.

Your browser likely doesn't use vulkan. Chrome for example has vulkan disabled by default. So the performance difference probably isn't from OpenGL vs Vulkan.

**Reply by BuslikDrev** [reply]
Установил принудительно OpenGL chrome://flags/#use-angle
1805x1015 35-50 fps
![image](https://github.com/opentk/opentk/assets/32706609/7e563149-3b24-4004-bf24-e7d752a6ca70)
![image](https://github.com/opentk/opentk/assets/32706609/9a75ab30-b887-436d-b3cb-22b763f1a449)

В моей программе 1805х1015 16 fps
![image](https://github.com/opentk/opentk/assets/32706609/4ae4657b-d4b3-4ffa-86fa-07e7f2f77f35)

Frag Шейдер назначал на куб,
```
		private static int[] IndexData = new int[] {
			 0,  1,  2,  2,  3,  0,
			 4,  5,  6,  6,  7,  4,
			 8,  9, 10, 10, 11,  8,
			12, 13, 14, 14, 15, 12,
			16, 17, 18, 18, 19, 16,
			20, 21, 22, 22, 23, 20,
		};
...

OpenTK.Graphics.OpenGL.GL.DrawArrays(OpenTK.Graphics.OpenGL.PrimitiveType.TriangleStrip, 0, IndexData.Length);
```

Сейчас попробовал установить ручное ограничение точек до 6.
```
OpenTK.Graphics.OpenGL.GL.DrawArrays(OpenTK.Graphics.OpenGL.PrimitiveType.TriangleStrip, 0, 6);
```
В результате вы правы, Спасибо, FPS выросло в 2 раза! Значит мне нужно создать плоский квадрат перед камерой.
![image](https://github.com/opentk/opentk/assets/32706609/b10045c5-20a4-4a11-b0e2-0b17efc6805c)
```
OpenTK.Graphics.OpenGL.GL.DrawArrays(OpenTK.Graphics.OpenGL.PrimitiveType.Quads, 0, 4);
```
1805x1015 50 fps
![image](https://github.com/opentk/opentk/assets/32706609/a8a3a1ac-f182-44ca-8acc-4d41c2e68086)

**Reply by NogginBops** [reply]
You could possibly get (slightly) better performance if you use what is usually called a "fullscreen triangle"
https://wallisc.github.io/rendering/2021/04/18/Fullscreen-Pass.html

**Reply by BuslikDrev** [reply]
Ускорение произошло на 1 FPS. Возможно, мне нужно отключить обработку пикселей в шейдере за пределами экрана, если это возможно. Если получится, опубликую код шейдера, ограничивающего отрисовку пикселей в треугольнике за пределами экрана.

1360x768
```
OpenTK.Graphics.OpenGL.GL.DrawArrays(OpenTK.Graphics.OpenGL.PrimitiveType.Quads, 0, 4);
```
![image](https://github.com/opentk/opentk/assets/32706609/2bbfb544-dd66-4c69-ba2b-f67f79a0710b)

```
OpenTK.Graphics.OpenGL.GL.DrawArrays(OpenTK.Graphics.OpenGL.PrimitiveType.Triangles, 0, 3);
```
![image](https://github.com/opentk/opentk/assets/32706609/4b4510f3-b883-4d46-b19e-0a054d5ab0b8)

**Reply by NogginBops** [reply]
The GPU will not process pixels outside of the viewport, so you don't really need to do anything there. The speedup from 95fps to 96fps is 0.1ms which is around the same values the article measures.

**Comment by BuslikDrev** [comment]
> OpenTK 3 работает на платформе .net, которая не имеет некоторых оптимизаций среды выполнения, присущих более поздним версиям .net. Некоторые математические функции в OpenTK 4 более оптимизированы. В остальном я не ожидаю существенной разницы в производительности между OpenTK 3 и OpenTK 4.

Thank you!
